lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -375,15 +375,15 @@ static void replaceLoad(LoadInst *LI, SILValue val, AllocStackInst *ASI) {
 
 static void replaceDestroy(DestroyAddrInst *DAI, SILValue NewValue) {
   SILFunction *F = DAI->getFunction();
+  auto Ty = DAI->getOperand()->getType();
 
-  assert(DAI->getOperand()->getType().isLoadable(*F) &&
+  assert(Ty.isLoadable(*F) &&
          "Unexpected promotion of address-only type!");
 
-  assert(NewValue && "Expected a value to release!");
+  assert(NewValue || (Ty.is<TupleType>() && Ty.getAs<TupleType>()->getNumElements() == 0));
```

Comment:
This assert makes sure that a value is actually written to the location before it is destroyed.
But this is not required for empty tuples.

Reply:
@eeckstein Disabling this assert means that an invalid SILValue is passed to `emitLoweredDestroyValue`. That is incorrect.

Reply:
emitLoweredDestroyValue is a no-op for an empty tuple.

Reply:
I get that it might happen to work, or might not. I'm just being picky about something trivial to make a point about interfaces. emitLoweredDestroyValue does not specify that it can take an invalid SILValue, and there's no single header-level implementation to check, so to anyone trying to understand this code it is incorrect at face value. It doesn't matter whether emitLoweredDestroyValue is actually going to emit an instruction or not. In general, if an interface can accept an invalid SILValue that needs to be specified in the interface. The same is true for accepting null pointers.

Reply:
ok, that's a good point